### PR TITLE
Blocks for generating user and project keys securely

### PR DESCRIPTION
**Pull request by worukan**
This attempt only includes a temporary key per user for the symmetric encryption. A key derived from the password will be introduced in another attempt soon.

### CODE REVIEW

File: dds_web/security/project_user_keys.py [hunk 1]
```diff
@@ -1,17 +1,128 @@
...
+    flask.current_app.logger.exception(exception)
+    raise exception
+
+
+def generate_user_key_pair(user):
     private_key = asymmetric.rsa.generate_private_key(public_exponent=65537, key_size=4096)
+    private_key_bytes = private_key.private_bytes(
+        serialization.Encoding.DER, serialization.PrivateFormat.PKCS8, serialization.NoEncryption()
+    )
+    encrypted_private_key = encrypt_user_private_key(user, private_key_bytes)
```
Comment (inaod568) [inline dds_web/security/project_user_keys.py:118]: You need to call the `private_key.public_key()` here (atm `del private_key` is before using it in the return), that's one of the reasons the tests fail atm.

**Comment by worukan** [comment]
This PR is ready for review.

As far as I can see from the dds_cli code, this PR should work with it. However, someone can test download/upload via cli and raise issues to be solved.

The implementation so far covers 

- project creation (with project keys),
- add research user to project (with setting up the project key for the user) and 
- create unit user (with setting up all project keys in the unit for the user).

Known issues that follow up

- Remove project
- Remove user
- Remove user from project
- Give project access to super admin?
- Write tests for the new blocks

File: dds_web/api/user.py [hunk 2]
```diff
@@ -213,6 +210,7 @@ def add_user_to_project(existing_user, project, role):
                     owner=owner,
                 )
             )
+            share_project_private_key_with_user(auth.current_user(), existing_user, project)
```
Comment (inaod568) [inline dds_web/api/user.py:213]: This should also be called if the researchuser has an account but is not connected to the project yet

File: dds_web/api/schemas/project_schemas.py [hunk 3]
```diff
@@ -153,24 +147,9 @@ def create_project(self, data, **kwargs):
...
-            # Add new project keys to table.
-            new_project_key = models.ProjectKeys(
-                key=aesgcm.encrypt(nonce=nonce, data=project_private_key, associated_data=aad),
-                nonce=nonce,
-            )
 
-            # Save
-            auth.current_user().project_keys.append(new_project_key)
-            new_project.project_keys.append(new_project_key)
+            generate_project_key_pair(current_user, new_project)
```
Comment (inaod568) [inline dds_web/api/schemas/project_schemas.py:151]: We need to add pproject key rows for all unitusers

File: dds_web/api/user.py (hunk 2 above)
Comment (worukan) [reply dds_web/api/user.py:213]: As I understand this function is executed when a ownership change happens as well. In that case there is no new project keys to share as the user already has access to it. It is only a title change of some sort not functional. If the function is not executed due to ownership change, it is to add an existing user to the project so line 213 will execute.

File: dds_web/security/project_user_keys.py [hunk 4]
```diff
@@ -1,17 +1,167 @@
...
+    encrypted_private_key = encrypt_project_private_key_for_user(user, private_key_bytes)
+    del private_key_bytes
+    del private_key
+    gc.collect()
+    project.public_key = public_key_bytes
+    project_user_key = models.ProjectUserKeys(
+        project_id=project.id, user_id=user.username, key=encrypted_private_key
+    )
+    user.project_user_keys.append(project_user_key)
+    project.project_user_keys.append(project_user_key)
```
[the remark this answers is not in the document]
Comment (worukan) [reply dds_web/security/project_user_keys.py:103]: I think this looks tidier than before as all details of db handling is done here.

File: dds_web/api/schemas/project_schemas.py (hunk 3 above)
Comment (worukan) [reply dds_web/api/schemas/project_schemas.py:151]: done in the last commit

**Comment by inaod568** [comment]
> This PR is ready for review.
> 
> As far as I can see from the dds_cli code, this PR should work with it. However, someone can test download/upload via cli and raise issues to be solved.
> 
> The implementation so far covers
> 
> * project creation (with project keys),
> * add research user to project (with setting up the project key for the user) and
> * create unit user (with setting up all project keys in the unit for the user).
> 
> Known issues that follow up
> 
> * Remove project
> * Remove user
> * Remove user from project
> * Give project access to super admin?
> * Write tests for the new blocks

Now also adding unit user keys at project creation.

**Comment by inaod568** [comment]
Lets fix the tests, then merge this PR and then we'll work on the follow up issues in other PRs.

File: dds_web/security/project_user_keys.py [hunk 5]
```diff
@@ -1,17 +1,168 @@
...
+    :param ciphertext: encrypted content
+    :param nonce: a unique value that has to be used only once to encrypt a given data with the given key
+    :param aad: Additional data that should be authenticated with the key, but is not encrypted. Can be None.
+    """
+    aesgcm = ciphers.aead.AESGCM(key)
+    return aesgcm.decrypt(nonce, ciphertext, aad)
+
+
+def encrypt_user_private_key(user, private_key):
+    user.temporary_key = ciphers.aead.AESGCM.generate_key(bit_length=256)
```
Comment (alneberg) [inline dds_web/security/project_user_keys.py:132]: Does this need a `db.session.commit` in order to be persisted?

Comment (inaod568) [reply dds_web/security/project_user_keys.py:132]: `encrypt_user_private_key` is called from `generate_user_key_pair` which currently is only called in `models.User.__init__`, and every time a user is created the `db.session.commit()` is called. So the commit isn't needed here at this point.

File: tests/conftest.py [hunk 6]
```diff
@@ -349,6 +330,21 @@ def client():
 
             db.session.commit()
 
+            generate_project_key_pair(users[2], units[0].projects[0])
+            generate_project_key_pair(users[2], units[0].projects[2])
```
Comment (alneberg) [inline tests/conftest.py:334]: This could be refactored but no reason for you to do it.

**Review by alneberg — APPROVED** [review]
Looks good, happy to merge this to continue building on it.

File: tests/conftest.py (hunk 6 above)
Comment (worukan) [reply tests/conftest.py:334]: Yes that's a piece of what I mentioned in the daily. It's there for now and better not in the long run.

**Comment by inaod568** [comment]
Waiting for the tests to be added, then I'm happy to approve and merge.

File: dds_web/api/schemas/user_schemas.py [hunk 7]
```diff
@@ -184,6 +181,8 @@ def make_user(self, data, **kwargs):
             new_user.is_admin = invite.role == "Unit Admin"
 
             invite.unit.users.append(new_user)
+            for project in invite.unit.projects:
+                share_project_private_key_with_user(auth.current_user(), new_user, project)
```
[the remark this answers is not in the document]
Comment (worukan) [reply dds_web/api/schemas/user_schemas.py]: Great catch! Removed it. I think we need to handle it at invite, but the implementation of that part has not started yet as we are waiting for @alneberg and @MatthiasZepper input.

**Review by inaod568 — APPROVED** [review]
Nice, looks good! 👏🏻

**Pull request merged**